agent/config/types.go

```diff
@@ -210,3 +219,42 @@ func (data *SensitiveRawMessage) UnmarshalJSON(jsonData []byte) error {
 	data.contents = json.RawMessage(jsonData)
 	return nil
 }
+
+// DefaultBool makes it possible to understand if a variable was set explicitly or relies on a default setting
+type DefaultBool int
+
+// Enabled is a convenience function for when consumers don't care if the value is implicit or explicit
+func (b DefaultBool) Enabled() bool {
+	return b == ExplicitlyEnabled || b == DefaultEnabled
+}
+
+func (b DefaultBool) MarshalJSON() ([]byte, error) {
+	switch b {
+	case ExplicitlyEnabled:
+		return json.Marshal(true)
+	case ExplicitlyDisabled:
+		return json.Marshal(false)
+	default:
+		return json.Marshal(nil)
+	}
+}
+
+func (b *DefaultBool) UnmarshalJSON(jsonData []byte) error {
+	jsonString := string(jsonData)
+	jsonBool, err := strconv.ParseBool(jsonString)
+
```

Comment:
This pattern appears to be a bit strange to me. 
* Why isn't the error check immediately performed ?
* Combining multiple checks within the switch is confusing ?

Reply:
> Why isn't the error check immediately performed ?

`ParseBool` returns an error for the `null` case, so there can be happy failure cases. But you're right, there is a more idiomatic way to express this.

> Combining multiple checks within the switch is confusing ?

[`switch` defaults to `switch(true)`. ](https://github.com/golang/go/wiki/Switch#missing-expression)  The idea was to put all of the conditions in one place. 

Given both of your comments, this may be better:

```golang

        // follow golang idiom closer with immediate nil check
        if err != nil && jsonString != "null" {
              return err
        }

        // I can swap this out with if / else 
	switch {
	case jsonString == "" || jsonString == "null":
		*b = DefaultEnabled
	case jsonBool:
		*b = ExplicitlyEnabled
        case !jsonBool:
		*b = ExplicitlyDisabled
	}
```

Reply:
I prefer the proposed approach and the `if/else` over `switch`